orttraining/orttraining/training_ops/cuda/activation/bias_gelu_grad_impl.cu

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cuda/activation/bias_gelu_grad_impl.h"
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/shared_inc/fast_divmod.h"
+#include "orttraining/training_ops/cpu/activation/gelu_computation_mode.h"
+#include "orttraining/training_ops/cuda/activation/gelu_grad_impl_common.cuh"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T, typename GeluComputationMode, int num_consecutive_elements_per_group, int num_groups_per_thread>
+__global__ void BiasGeluGradDxKernel(
+    CUDA_LONG input_size, fast_divmod bias_size_fdm,
+    const T* dY, const T* X, const T* B, T* dX) {
+  const auto& num_threads_per_block = blockDim.x;
+  const int& bias_size = bias_size_fdm.d_;
+  CUDA_LONG base_idx =
+      num_consecutive_elements_per_group * num_groups_per_thread * num_threads_per_block * blockIdx.x +
+      num_consecutive_elements_per_group * threadIdx.x;
+
+#pragma unroll
+  for (int i = 0; i < num_groups_per_thread; ++i) {
+    if (base_idx < input_size) {
+      int q, r;
+      bias_size_fdm.divmod(base_idx, q, r);
+      const int& base_bias_idx = r;
+
+#pragma unroll
+      for (int element_idx = 0; element_idx < num_consecutive_elements_per_group; ++element_idx) {
```

Comment:
>++element_idx [](start = 82, length = 13)

This accessing 4 elements consecutively... 

According to Weixing's previous experiment, this would be slower than accessing 4 elements with a stride

Reply:
made some measurements with different configurations of num_groups_per_thread x num_consecutive_elements_per_group:
```
Type  Time(%)      Time     Calls       Avg       Min       Max
 4x4    2.54%  97.423ms       250  389.69us  108.48us  414.33us
 8x2    2.68%  102.48ms       250  409.93us  112.48us  435.10us
 4x1    2.72%  104.16ms       250  416.64us  109.79us  462.88us
16x1    2.91%  112.02ms       250  448.07us  119.01us  475.32us
 2x8    2.96%  113.23ms       250  452.91us  125.06us  484.28us
1x16    5.68%  223.82ms       250  895.29us  248.32us  1.0040ms
```
looks like 4x4 is a reasonable choice

---
In reply to: [441112899](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r441112899) [](ancestors = 441112899)

Reply:
refactored this a bit

---
In reply to: [441900620](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r441900620) [](ancestors = 441900620,441112899)

Reply:
Please share the result on the PR top description. 

---
In reply to: [442609897](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r442609897) [](ancestors = 442609897,441900620,441112899)

Reply:
ran another test against the baseline and put the results in the description. ran the second test on a different machine so the numbers look different, but the relative comparison should be valid

---
In reply to: [443771031](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r443771031) [](ancestors = 443771031,442609897,441900620,441112899)